src/bokeh/core/property/instance.py

```diff
@@ -111,6 +114,12 @@ def instance_type(self) -> type[T]:
 
     def validate(self, value: Any, detail: bool = True) -> None:
         super().validate(value, detail)
+        if not nw.dependencies.is_pandas_dataframe(value):
+            # Try converting to Narwhals so that any dataframe supported by Narwhals
+            # can be supported by ColumnDataSource. We don't perform this conversion
+            # for pandas so that code written using `Object('pandas.Dataframe)` can
+            # keep working (backwards-compatibility).
+            value = nw.from_native(value, pass_through=True)
```

Comment:
I only see two places this seems to be used
```
src/bokeh/plotting/_figure.py
818:    Object("pandas.Series"),
819:    Object("pandas.core.groupby.GroupBy"),

src/bokeh/models/sources.py
210:        Object("pandas.DataFrame"), lambda x: ColumnDataSource._data_from_df(x),
212:        Object("pandas.core.groupby.GroupBy"), lambda x: ColumnDataSource._data_from_groupby(x),
```

Not sure what your interest level is but TBH prefer to fix this by deprecating or removing the types in`property/pd.py` and adding "generic" `DataFrame` and `GroupBy` [1] properties backed by `narwhals` validation. Then update those couple of places to use those instead. 

[1] Maybe `EagerDataFrame` e.g. if we want to be very specific

Reply:
Sure, thanks @bryevdv !

OK glad to hear this 😄 It simplifies the code

I've unified validation here under `Object('narwhals.DataFrame')` / `Object('narwhals.Series')`, introduced `EagerDataFrame`, and deprecated `PandasDataFrame`

The part I currently can't address is `GroupBy`. Narwhals supports writing dataframe-agnostic code when starting from a DataFrame or Series, but not when starting from a GroupBy object (and I think it'll probably stay out of scope - different libraries implement group-by very differently, and accepting a GroupBy object isn't a common use-case). Furthermore, for PyArrow, I don't think there's a way to mimic pandas' `GroupBy.describe` using public methods when starting from just a [pyarrow.TableGroupBy](https://arrow.apache.org/docs/python/generated/pyarrow.TableGroupBy.html) object

So, I'd like to suggest that GroupBy remains a pandas-only thing (for backwards-compatibility), but that DataFrame becomes library-agnostic - would this work for you?

> Not sure what your interest level is

I'm happy to work on this feature, but please do feel free to add commits if that's the easiest way to move things along 🙏 Appreciate your help here!